### PR TITLE
Document the project policy on mixing build systems. [skip ci]

### PR DESCRIPTION
**Pull request by jpakkane**

### CODE REVIEW

File: docs/markdown/Mixing-build-systems.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# Meson's policy on mixing multiple build systems in one build directory
+
+Meson has been designed with the principle that all dependencies are
+either provided by "the platform" via a mechanism such as Pkg-Config
+or that they are built as Meson subprojects under the main
+project. There are several projects that would like to mix build
+systems, that is, build dependencies in the same build directory in
+the same build directory by having one build system call the other.
```
Comment (jon-turney) [inline docs/markdown/Mixing-build-systems.md]: "in the same build directory in the same build directory" (duplicated)

File: docs/markdown/Mixing-build-systems.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+This page lists the Meson project's stance on mixing build system. The
+tl/dr version is that while we do provide some functionality for this
+use case, it only works for simple cases. Anything more complex can
+not be made reliable. Thus the support for these use cases is not
+guaranteed to work (in fact it is almost certainly guaranteed to *not*
```
Comment (mensinda) [inline docs/markdown/Mixing-build-systems.md]: I agree with making our policy on mixing build systems clear, but I think this part is a bit too harsh. Would something like `Thus the support for these use cases is unlikely to work.` also work?

Also maybe consider changing `it only works for simple cases` to `it only somewhat reliably works for simple cases`. IMHO if these mechanisms can't at least be classified to work "somewhat reliable" for small/simple projects (which *does* reflect the current status for CMake) then there is no point in having them :)

File: docs/markdown/Mixing-build-systems.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+work), and even if it works today there are no guarantees that it will
+keep working in any future version.
+
+## The definition of "buils system mixing"
```
Comment (tp-m) [inline docs/markdown/Mixing-build-systems.md]: Typo ("buils")

**Comment by xclaesse** [comment]
I'm not sure if this means cargo module as proposed in https://github.com/mesonbuild/meson/pull/2617 would be accepted or not.

**Comment by nirbheek** [comment]
> I'm not sure if this means cargo module as proposed in #2617 would be accepted or not.

I would interpret this as meaning "go ahead and merge it", because it meets all these requirements.

File: docs/markdown/Mixing-build-systems.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
+# Meson's policy on mixing multiple build systems in one build directory
+
+Meson has been designed with the principle that all dependencies are
+either provided by "the platform" via a mechanism such as Pkg-Config
+or that they are built as Meson subprojects under the main
+project. There are several projects that would like to mix build
+systems, that is, build dependencies in the same build directory as
+the other build system by having one build system call the other. The
+build diectories do not necessarily need to be inside each other, but
```
Comment (nirbheek) [inline docs/markdown/Mixing-build-systems.md]: *build directories

File: docs/markdown/Mixing-build-systems.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+that is the common case.
+
+This page lists the Meson project's stance on mixing build system. The
+tl/dr version is that while we do provide some functionality for this
+use case, it only works for simple cases. Anything more complex can
+not be made reliable and trying to do that would burden Meson
+developers with an effectively infinite maintenance burder. Thus these
```
Comment (nirbheek) [inline docs/markdown/Mixing-build-systems.md]: *burden

File: docs/markdown/Mixing-build-systems.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+Meson has been designed with the principle that all dependencies are
+either provided by "the platform" via a mechanism such as Pkg-Config
+or that they are built as Meson subprojects under the main
+project. There are several projects that would like to mix build
+systems, that is, build dependencies in the same build directory as
+the other build system by having one build system call the other. The
+build diectories do not necessarily need to be inside each other, but
+that is the common case.
+
+This page lists the Meson project's stance on mixing build system. The
```
Comment (nirbheek) [inline docs/markdown/Mixing-build-systems.md]: *build systems

**Comment by jpakkane** [comment]
The point of this PR was to make it possible to merge things like #2617. It does not mean that they will be automatically and blindly merged, though. They must be reviewed as usual and in addition every piece of functionality (mostly modules) that falls under this rule **must** start its documentation by referring to this guide (like the CMake module does in this commit).

**Pull request merged**